### ISSUE TITLE
Add a configuration to serve a local directory with a static handler

### ISSUE THREAD

**Issue by ia3andy** [quarkusio/quarkus#39968]
### Description

Currently we are suggesting to setup a custom route manually:
https://quarkus.io/guides/http-reference#from-a-local-directory

This would be more elegant and cleaner if it was done in the Quarkus Http using a config

**Comment by geoand** [comment]
Definitely +1 for this.

**Comment by ia3andy** [comment]
@geoand do you want to add it?

**Comment by geoand** [comment]
I am pretty sure anyone could do it 😉

**Comment by geoand** [comment]
@ia3andy here you go: https://github.com/quarkusio/quarkus/pull/39993

**Comment by cescoffier** [comment]
I commented on the PR - do NOT add multiple static handlers.

All the static handlers rely on the same file resolver, which contains a cache. If you have concurrent access (which you can do as soon as your number of event loops is greater than 1), BOOM. 

We would have to verify that the directories do not overlap and that file names are unique (otherwise, the cache won't work).

**Comment by geoand** [comment]
@cescoffier then our docs which tell people to add one are also wrong.

**Comment by cescoffier** [comment]
Yes and no - it can work, but it may not either (in the sense it could leak or corrupt). 

If your path do not overlap and file names are unique, it is going to work (I still need to check ranges).

**Comment by geoand** [comment]
> If your path do not overlap

You mean that each handler should serve a unique root, correct?

 >  and file names are unique

What does this mean exactly? That different instances of StaticHandler can't serve two totally different files that happen to have the same name?

**Comment by cescoffier** [comment]
Yes, different roots without overlap (including symbolic links and so on). 
And yes, the file names must be unique because of the cache key. 

See https://github.com/eclipse-vertx/vert.x/blob/4.5.7/src/main/java/io/vertx/core/file/impl/FileResolverImpl.java#L262 
and https://github.com/eclipse-vertx/vert.x/blob/4.5.7/src/main/java/io/vertx/core/file/impl/FileCache.java#L177.

Note the (on purpose) lack of synchronization (and now you understand why I am concerned)

**Comment by geoand** [comment]
Yes, I see your point. 

I am going to close my PR since this needs to be handled properly in Vertx before we do anything in Quarkus to make things easier.

**Comment by cescoffier** [comment]
Note that if we want to support only resources from the file system (not from the classloader), things are way easier (and a custom, optimized static handler can be created in a few minutes)

**Comment by geoand** [comment]
Right, the way I see it, this issue is _only_ about file system resources, not classpath resources.

**Comment by ia3andy** [comment]
I think we should remove this section from the documentation until this get fixed

**Comment by mcruzdev** [comment]
Hi @ia3andy, can you assign it to me too?

**Comment by cescoffier** [comment]
@geoand No, I think it should be in Quarkus as we can improve things at build time.

**Comment by geoand** [comment]
Fair enough

**Comment by ia3andy** [comment]
@cescoffier wouldn't that be possible in Vertx to cache with a relative path instead of just the file name?

**Comment by cescoffier** [comment]
@ia3andy check the Vertx Cache implementation, my memory is a bit rusty.

**Comment by cescoffier** [comment]
@ia3andy did you check on the Vert.x side? 

Pretty sure that `..` is forbidden to avoid leaving the served directory (security issue)

**Comment by ia3andy** [comment]
Maybe we could do this:
```java
// for each files in the configured static dir:
String link = PathUtils.join(config.rootPath(), PathUtils.toUnixPath(staticDir.relativize(staticFile)));
 staticResourcesProducer.produce(new GeneratedStaticResourceBuildItem(link, staticFile));
```

**Comment by cescoffier** [comment]
@ia3andy are you planning to contribute this?

**Comment by mcruzdev** [comment]
Hi @cescoffier this [comment](https://github.com/quarkusio/quarkus/issues/39968#issuecomment-2049005212) was solved by Vertx? I am confused to start due to some previous comments.

**Comment by ia3andy** [comment]
> Yes, it would require some changes in vert.x as at the moment there is a single file resolver.

not if we use the GeneratedStaticResource solutioon?

**Comment by mcruzdev** [comment]
> Maybe we could do this (now that we have GeneratedStaticResourceBuildItem):
> 
> // for each files in the configured static dir:
> String link = PathUtils.toUnixPath(staticDir.relativize(staticFile)); // Make sure we convert \ to / and use the relative path from the directory as link
>  staticResourcesProducer.produce(new GeneratedStaticResourceBuildItem(link, staticFile));

ok

**Comment by cescoffier** [comment]
Be careful that this approach does not lead to path traversal vulnerabilities.

**Comment by ia3andy** [comment]
@cescoffier only what's detected at build time is marked to be served, so we should be fine..

**Comment by mathias82** [comment]
Hello!
I've submitted a PR implementing this enhancement:
Adds a configurable local static directory served through Vert.x HTTP.
PR: https://github.com/quarkusio/quarkus/pull/51186

The local static resources are served using Vert.x StaticHandler, which normalizes request paths and prevents directory traversal outside the configured root directory. The feature is disabled by default and is intended for serving public static assets. 
Users should avoid pointing it to directories containing sensitive data.

Let me know if any changes are required.

**Comment by ia3andy** [comment]
@mathias82 https://github.com/quarkusio/quarkus/issues/39968#issuecomment-2048981605

**Comment by mathias82** [comment]
Thanks i missed that! @ia3andy 

You're absolutely right, creating multiple StaticHandlers would create
multiple FileResolvers and break caching under concurrency.

I'll update the PR so that only a single StaticHandler instance is created
and reused for the configured local directory.  
This ensures proper caching behaviour and avoids race conditions.

Let me know if you prefer it to be stored in a static field, injected via
recorder, or added as part of the existing static handler initialization.

**Comment by ia3andy** [comment]
@mathias82 as mentioned before, you can use `GeneratedStaticResourceBuildItem` (https://github.com/quarkusio/quarkus/blob/main/extensions/vertx-http/deployment-spi/src/main/java/io/quarkus/vertx/http/deployment/spi/GeneratedStaticResourceBuildItem.java) instead

**Comment by mathias82** [comment]
Thanks for the help @ia3andy can you take a look the pr?  

I removed the Vert.x `StaticHandler` approach and completely switched to using
`GeneratedStaticResourceBuildItem`, as you suggested.

This keeps the implementation aligned with existing static resource mechanisms
and avoids the concurrency issues you mentioned.

Let me know if you'd like any further adjustments!
